contracts/EIP712CrossChain.sol

```diff
@@ -0,0 +1,85 @@
+// SPDX-License-Identifier: GPL-3.0-only
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+
+/**
+ * @dev https://eips.ethereum.org/EIPS/eip-712[EIP 712] is a standard for hashing and signing of typed structured data.
+ *
+ * This contract is based on OpenZeppelin's implementation:
+ * https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v4.8.0/contracts/utils/cryptography/EIP712.sol
+ *
+ * NOTE: Modified version that allows to build a domain separator that relies on a different chain id than the chain this
+ * contract is deployed to. An example use case we want to support is:
+ * - User A signs a message on chain with id = 1
+ * - User B executes a method by verifying user A's EIP-712 compliant signature on a chain with id != 1
+ */
+abstract contract EIP712CrossChain {
+    /* solhint-disable var-name-mixedcase */
+    bytes32 private immutable _HASHED_NAME;
+    bytes32 private immutable _HASHED_VERSION;
+    bytes32 private immutable _TYPE_HASH;
+    // NOTE: We use a hardcoded address to allow cross-chain verification
+    address public immutable _VERIFYING_CONTRACT = 0xCcCCccccCCCCcCCCCCCcCcCccCcCCCcCcccccccC;
```

Comment:
Would we need to change this in production? Isn't this contract the destination SpokePool's address?

Reply:
This is again the same problem as with the chain id. Because this value is part of the domain separator, we either need to hardcode it across all chains / spoke pool contracts OR make it also a parameter. I decided to hardcode it because:
1. The spec defines this only with the keyword _may_, so my interpretation is that this is optional. See https://eips.ethereum.org/EIPS/eip-712#definition-of-domainseparator
2. If this would be a parameter, then the function signatures of the methods `speedUpDeposit` and `fillRelayWithUpdatedFee` would need to change. Because we would need to somehow pass the address of the respective spoke pool (same value for corresponding chains)?

Reply:
hmmm this is an interesting point @dohaki . So one way around this is to just hardcode the verifying contract so that its essentially ignored. Does that break wallet clients ability to parse this message? I just don't want them to reject it because the `verifyingContract` isn't actually the contract that is verifying it

Reply:
As discussed offline, it seems that some of the fields like `verifyingContract` are optional in the EIP so we don't need to use that one. I still think we should keep `chainId = originChainId`. If we remove `verifyingContract` i don't think we expose ourselves to any replay attacks where someone submits the signature to a different spoke pool on the `originChain`. The SpokePools now allow fills to be paused including the `fillRelayWithUpdatedFee` so that will revert on deprecated SpokePools. Additionally, even if the `fillRelayWithUpdatedFee` succeeds, the relayer will note receive a refund if the SpokePool isn't connected to the HubPool anymore.